stl/inc/algorithm

```diff
@@ -785,6 +785,15 @@ namespace ranges {
         template <class _It1, class _Se1, class _It2, class _Se2, class _Pr, class _Pj1, class _Pj2>
         _NODISCARD static constexpr bool _Equal_4(
             _It1 _First1, _Se1 _Last1, _It2 _First2, _Se2 _Last2, _Pr _Pred, _Pj1 _Proj1, _Pj2 _Proj2) {
+            if constexpr (_Equal_memcmp_is_safe<_It1, _It2, _Pr> && same_as<_Pj1, identity> //
+                          && same_as<_Pj2, identity> //
+                          && (same_as<_Se1, unreachable_sentinel_t> || same_as<_Se2, unreachable_sentinel_t>) ) {
+                if constexpr (same_as<_Se1, _Se2>) {

```

Comment:
I think it would be cleaner, if we would add that as a first case, aka
```cpp
if constexpr (is_same_v<_Se1, unreachable_sentinel_t> && is_same_v<_Se2, unreachable_sentinel_t>) ) {
    _STATIC_ASSERT(_Always_false<_It1>,  "Tried to compare ranges with both sentinels being unreachable");
} else if constexpr (_Equal_memcmp_is_safe<_It1, _It2, _Pr> ...
```

Note that we should also turn the `_STL_ASSERT` into a `_STATIC_ASSERT` with the `_Always_false` helper

Reply:
There might some use cases with custom iterators that do weird things that might want to call `ranges::equal` with two unreachable sentinels (maybe for some side effects), the only thing I'm sure about is that there is no such case if `memcmp` could be used.

Also AFAIK the standard doesn't say that if both sentinels are unreachable the code doesn't have to compile so it can't use `_STATIC_ASSERT`.